casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -173,54 +173,68 @@ object InterpreterUtil {
-          for {
-            parentsMetadata    <- parents.toList.traverse(b => dag.lookup(b.blockHash).map(_.get))
-            indexedParents     = parentsMetadata.toVector
-            uncommonAncestors  <- DagOperations.uncommonAncestors[F](indexedParents, dag)
-            initParentMetadata <- dag.lookup(initParent.blockHash)
-            initIndex          = indexedParents.indexOf(initParentMetadata)
-            //filter out blocks that already included by starting from the chosen initParent
-            blocksToApply = uncommonAncestors
-              .filterNot { case (_, set) => set.contains(initIndex) }
-              .keys
-              .toVector
-              .sorted //ensure blocks to apply is topologically sorted to maintain any causal dependencies
-            maybeBlocks <- blocksToApply.traverse(b => BlockStore[F].get(b.blockHash))
-            _           = assert(maybeBlocks.forall(_.isDefined))
-            blocks      = maybeBlocks.flatten
-            deploys     = blocks.flatMap(_.getBody.deploys.flatMap(ProcessedDeployUtil.toInternal))
-          } yield
-            runtimeManager
-              .replayComputeState(initStateHash, deploys, time)
-              .runSyncUnsafe(Duration.Inf) match {
-              case result @ Right(_) => result.leftCast[Throwable]
-              case Left((_, status)) =>
-                val parentHashes = parents.map(p => Base16.encode(p.blockHash.toByteArray).take(8))
-                Left(
-                  new Exception(
-                    s"Failed status while computing post state of $parentHashes: $status"
-                  )
-                )
-            }
-        }
+      case (_, initStateHash) +: _ =>
+        computeMultiParentsPostState[F](
+          parents,
+          dag,
+          runtimeManager,
+          time,
+          initStateHash
+        )
     }
   }
 
+  // In the case of multiple parents we need to apply all of the deploys that have been
+  // made in all of the branches of the DAG being merged. This is done by computing uncommon ancestors
+  // and applying the deploys in those blocks on top of the initial parent.
+  private def computeMultiParentsPostState[F[_]: Monad: BlockStore](
+      parents: Seq[BlockMessage],
+      dag: BlockDagRepresentation[F],
+      runtimeManager: RuntimeManager,
+      time: Option[Long],
+      initStateHash: StateHash
+  )(implicit scheduler: Scheduler): F[Either[Throwable, StateHash]] =
+    for {
+      parentsMetadata <- parents.toList.traverse(b => dag.lookup(b.blockHash).map(_.get))
+      ordering        <- dag.deriveOrdering(0L) // TODO: Replace with an actual starting number
+      blockHashesToApply <- {
+        implicit val o: Ordering[BlockMetadata] = ordering
+        for {
+          uncommonAncestors          <- DagOperations.uncommonAncestors[F](parentsMetadata.toVector, dag)
+          ancestorsOfInitParentIndex = 0
```

Comment:
This doesn't change between iterations so can be moved to a constant at the beginning of the method

Reply:
There is no iterations going on here.